### PR TITLE
[FIX] UTF-8 Decode error, singleton error in check methods, serializ

### PR DESCRIPTION
**Pull request by CasVissers-360ERP**
[FIX] UTF-8 Decode error, singleton error in check methods, serialize exception correctly

Will provide tracebacks for all fixes when required.

### CODE REVIEW

**Review by Ramil-Mukhametzyanov — COMMENTED** [review]
Hello @CasVissers 
Thank you for the pull-request. Could you also send us error logs.

File: openapi/controllers/apijsonrequest.py [hunk 1]
```diff
@@ -39,16 +40,16 @@ def __init__(self, *args):
         if jsonp and self.httprequest.method == 'POST':
             # jsonp 2 steps step1 POST: save call
             def handler():
-                self.session['jsonp_request_%s' % (request_id,)] = self.httprequest.form['r']
+                self.session['jsonp_request_%s' % (request_id,)] = self.httprequest.form['rb']
```
Comment (yelizariev) [inline openapi/controllers/apijsonrequest.py:43]: This part of code was made by @blaggacao and I don't know about it much. Could you explain a little bit where ``'r'`` and ``'rb'`` are come from?

**Comment by CasVissers-360ERP** [comment]
Singleton Traceback:
```
Error:
Odoo Server Error

Traceback (most recent call last):
  File "/Users/brahoo/PycharmProjects/odoo/odoo/models.py", line 4981, in ensure_one
    _id, = self._ids
ValueError: too many values to unpack (expected 1)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/brahoo/PycharmProjects/odoo/odoo/http.py", line 619, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/Users/brahoo/PycharmProjects/odoo/odoo/http.py", line 309, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/Users/brahoo/PycharmProjects/odoo/odoo/tools/pycompat.py", line 14, in reraise
    raise value
  File "/Users/brahoo/PycharmProjects/odoo/odoo/http.py", line 664, in dispatch
    result = self._call_function(**self.params)
  File "/Users/brahoo/PycharmProjects/odoo/odoo/http.py", line 345, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/Users/brahoo/PycharmProjects/odoo/odoo/service/model.py", line 93, in wrapper
    return f(dbname, *args, **kwargs)
  File "/Users/brahoo/PycharmProjects/odoo/odoo/http.py", line 338, in checked_call
    result = self.endpoint(*a, **kw)
  File "/Users/brahoo/PycharmProjects/odoo/odoo/http.py", line 909, in __call__
    return self.method(*args, **kw)
  File "/Users/brahoo/PycharmProjects/odoo/odoo/http.py", line 510, in response_wrap
    response = f(*args, **kw)
  File "/Users/brahoo/PycharmProjects/odoo/addons/web/controllers/main.py", line 1319, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "/Users/brahoo/PycharmProjects/odoo/addons/web/controllers/main.py", line 1311, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/Users/brahoo/PycharmProjects/odoo/odoo/api.py", line 395, in call_kw
    result = _call_kw_multi(method, model, args, kwargs)
  File "/Users/brahoo/PycharmProjects/odoo/odoo/api.py", line 382, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/Users/brahoo/PycharmProjects/customer/community/openapi/models/openapi_namespace.py", line 78, in write
    return super(Namespace, self).write(vals)
  File "/Users/brahoo/PycharmProjects/odoo/odoo/models.py", line 3532, in write
    field.write(self, vals[fname])
  File "/Users/brahoo/PycharmProjects/odoo/odoo/fields.py", line 2745, in write
    return self.write_batch([(records, value)])
  File "/Users/brahoo/PycharmProjects/odoo/odoo/fields.py", line 2766, in write_batch
    return self.write_real(records_commands_list, create)
  File "/Users/brahoo/PycharmProjects/odoo/odoo/fields.py", line 2930, in write_real
    flush()
  File "/Users/brahoo/PycharmProjects/odoo/odoo/fields.py", line 2894, in flush
    comodel.create(to_create)
  File "<decorator-gen-3>", line 2, in create
  File "/Users/brahoo/PycharmProjects/odoo/odoo/api.py", line 344, in _model_create_multi
    return create(self, arg)
  File "/Users/brahoo/PycharmProjects/odoo/odoo/models.py", line 3732, in create
    records = self._create(data_list)
  File "/Users/brahoo/PycharmProjects/odoo/odoo/models.py", line 3885, in _create
    records._validate_fields(name for data in data_list for name in data['stored'])
  File "/Users/brahoo/PycharmProjects/odoo/odoo/models.py", line 1167, in _validate_fields
    check(self)
  File "/Users/brahoo/PycharmProjects/customer/community/openapi/models/openapi_access.py", line 127, in _check_methods
    self.api_create,
  File "/Users/brahoo/PycharmProjects/odoo/odoo/fields.py", line 966, in __get__
    record.ensure_one()
  File "/Users/brahoo/PycharmProjects/odoo/odoo/models.py", line 4984, in ensure_one
    raise ValueError("Expected singleton: %s" % self)
ValueError: Expected singleton: openapi.access(8, 9)
```

**Comment by CasVissers-360ERP** [comment]
Serialize Traceback:
```
2020-01-31 08:24:19,637 9095 ERROR customer werkzeug: Error on request:
Traceback (most recent call last):
  File "/Users/brahoo/odoo13-env/lib/python3.7/site-packages/werkzeug/serving.py", line 270, in run_wsgi
    execute(self.server.app)
  File "/Users/brahoo/odoo13-env/lib/python3.7/site-packages/werkzeug/serving.py", line 258, in execute
    application_iter = app(environ, start_response)
  File "/Users/brahoo/PycharmProjects/odoo/odoo/service/server.py", line 439, in app
    return self.app(e, s)
  File "/Users/brahoo/PycharmProjects/odoo/odoo/service/wsgi_server.py", line 142, in application
    return application_unproxied(environ, start_response)
  File "/Users/brahoo/PycharmProjects/odoo/odoo/service/wsgi_server.py", line 117, in application_unproxied
    result = odoo.http.root(environ, start_response)
  File "/Users/brahoo/PycharmProjects/odoo/odoo/http.py", line 1281, in __call__
    return self.dispatch(environ, start_response)
  File "/Users/brahoo/PycharmProjects/odoo/odoo/http.py", line 1251, in __call__
    return self.app(environ, start_wrapped)
  File "/Users/brahoo/odoo13-env/lib/python3.7/site-packages/werkzeug/wsgi.py", line 766, in __call__
    return self.app(environ, start_response)
  File "/Users/brahoo/PycharmProjects/odoo/odoo/http.py", line 1450, in dispatch
    result = ir_http._dispatch()
  File "/Users/brahoo/PycharmProjects/odoo/addons/auth_signup/models/ir_http.py", line 19, in _dispatch
    return super(Http, cls)._dispatch()
  File "/Users/brahoo/PycharmProjects/odoo/addons/web_editor/models/ir_http.py", line 22, in _dispatch
    return super(IrHttp, cls)._dispatch()
  File "/Users/brahoo/PycharmProjects/odoo/addons/utm/models/ir_http.py", line 29, in _dispatch
    response = super(IrHttp, cls)._dispatch()
  File "/Users/brahoo/PycharmProjects/odoo/addons/http_routing/models/ir_http.py", line 516, in _dispatch
    result = super(IrHttp, cls)._dispatch()
  File "/Users/brahoo/PycharmProjects/odoo/odoo/addons/base/models/ir_http.py", line 233, in _dispatch
    return cls._handle_exception(e)
  File "/Users/brahoo/PycharmProjects/odoo/addons/utm/models/ir_http.py", line 34, in _handle_exception
    response = super(IrHttp, cls)._handle_exception(exc)
  File "/Users/brahoo/PycharmProjects/odoo/addons/http_routing/models/ir_http.py", line 607, in _handle_exception
    return super(IrHttp, cls)._handle_exception(exception)
  File "/Users/brahoo/PycharmProjects/odoo/odoo/addons/base/models/ir_http.py", line 201, in _handle_exception
    return request._handle_exception(exception)
  File "/Users/brahoo/PycharmProjects/customer/community/openapi/controllers/apijsonrequest.py", line 99, in _handle_exception
    'data': request.registry['ir.http'].serialize_exception(exception)
AttributeError: type object 'ir.http' has no attribute 'serialize_exception' - - -

```

**Comment by CasVissers-360ERP** [comment]
UTF-8 decode ('rb'):
```
Traceback (most recent call last):
  File "/Users/brahoo/PycharmProjects/customer/community/openapi/controllers/apijsonrequest.py", line 92, in _handle_exception
    return super(ApiJsonRequest, self)._handle_exception(exception)
  File "/Users/brahoo/PycharmProjects/odoo/odoo/http.py", line 309, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/Users/brahoo/PycharmProjects/odoo/odoo/tools/pycompat.py", line 14, in reraise
    raise value
  File "/Users/brahoo/PycharmProjects/customer/community/openapi/controllers/apijsonrequest.py", line 135, in dispatch
    result = self._call_function(**self.params)
  File "/Users/brahoo/PycharmProjects/odoo/odoo/http.py", line 345, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/Users/brahoo/PycharmProjects/odoo/odoo/service/model.py", line 93, in wrapper
    return f(dbname, *args, **kwargs)
  File "/Users/brahoo/PycharmProjects/odoo/odoo/http.py", line 338, in checked_call
    result = self.endpoint(*a, **kw)
  File "/Users/brahoo/PycharmProjects/odoo/odoo/http.py", line 909, in __call__
    return self.method(*args, **kw)
  File "/Users/brahoo/PycharmProjects/customer/community/openapi/controllers/apijsonrequest.py", line 170, in response_wrap
    response = f(*args, **kw)
  File "/Users/brahoo/PycharmProjects/customer/community/openapi/controllers/pinguin.py", line 326, in controller_method_wrapper
    authenticated_user = authenticate_token_for_user(user_token)
  File "/Users/brahoo/PycharmProjects/customer/community/openapi/controllers/pinguin.py", line 145, in authenticate_token_for_user
    user = request.env['res.users'].sudo().search([('openapi_token', '=', token)])
  File "/Users/brahoo/PycharmProjects/odoo/odoo/models.py", line 1609, in search
    res = self._search(args, offset=offset, limit=limit, order=order, count=count)
  File "/Users/brahoo/PycharmProjects/odoo/odoo/addons/base/models/res_users.py", line 470, in _search
    access_rights_uid=access_rights_uid)
  File "/Users/brahoo/PycharmProjects/odoo/odoo/models.py", line 4363, in _search
    query = self._where_calc(args)
  File "/Users/brahoo/PycharmProjects/odoo/odoo/models.py", line 4103, in _where_calc
    where_clause, where_params = e.to_sql()
  File "/Users/brahoo/PycharmProjects/odoo/odoo/osv/expression.py", line 1287, in to_sql
    q, ps = self.__leaf_to_sql(leaf)
  File "/Users/brahoo/PycharmProjects/odoo/odoo/osv/expression.py", line 1276, in __leaf_to_sql
    params = [field.convert_to_column(right, model, validate=False)]
  File "/Users/brahoo/PycharmProjects/odoo/odoo/fields.py", line 1507, in convert_to_column
    return pycompat.to_text(value)[:self.size]
  File "/Users/brahoo/PycharmProjects/odoo/odoo/tools/pycompat.py", line 46, in to_text
    return source.decode('utf-8')
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xed in position 0: invalid continuation byte
2020-01-31 08:32:53,346 10318 ERROR customer odoo.addons.openapi.controllers.apijsonrequest: Exception during JSON request handling. 
Traceback (most recent call last):
  File "/Users/brahoo/PycharmProjects/customer/community/openapi/controllers/apijsonrequest.py", line 92, in _handle_exception
    return super(ApiJsonRequest, self)._handle_exception(exception)
  File "/Users/brahoo/PycharmProjects/odoo/odoo/http.py", line 309, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/Users/brahoo/PycharmProjects/odoo/odoo/tools/pycompat.py", line 14, in reraise
    raise value
  File "/Users/brahoo/PycharmProjects/customer/community/openapi/controllers/apijsonrequest.py", line 135, in dispatch
    result = self._call_function(**self.params)
  File "/Users/brahoo/PycharmProjects/odoo/odoo/http.py", line 345, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/Users/brahoo/PycharmProjects/odoo/odoo/service/model.py", line 93, in wrapper
    return f(dbname, *args, **kwargs)
  File "/Users/brahoo/PycharmProjects/odoo/odoo/http.py", line 338, in checked_call
    result = self.endpoint(*a, **kw)
  File "/Users/brahoo/PycharmProjects/odoo/odoo/http.py", line 909, in __call__
    return self.method(*args, **kw)
  File "/Users/brahoo/PycharmProjects/customer/community/openapi/controllers/apijsonrequest.py", line 170, in response_wrap
    response = f(*args, **kw)
  File "/Users/brahoo/PycharmProjects/customer/community/openapi/controllers/pinguin.py", line 326, in controller_method_wrapper
    authenticated_user = authenticate_token_for_user(user_token)
  File "/Users/brahoo/PycharmProjects/customer/community/openapi/controllers/pinguin.py", line 145, in authenticate_token_for_user
    user = request.env['res.users'].sudo().search([('openapi_token', '=', token)])
  File "/Users/brahoo/PycharmProjects/odoo/odoo/models.py", line 1609, in search
    res = self._search(args, offset=offset, limit=limit, order=order, count=count)
  File "/Users/brahoo/PycharmProjects/odoo/odoo/addons/base/models/res_users.py", line 470, in _search
    access_rights_uid=access_rights_uid)
  File "/Users/brahoo/PycharmProjects/odoo/odoo/models.py", line 4363, in _search
    query = self._where_calc(args)
  File "/Users/brahoo/PycharmProjects/odoo/odoo/models.py", line 4103, in _where_calc
    where_clause, where_params = e.to_sql()
  File "/Users/brahoo/PycharmProjects/odoo/odoo/osv/expression.py", line 1287, in to_sql
    q, ps = self.__leaf_to_sql(leaf)
  File "/Users/brahoo/PycharmProjects/odoo/odoo/osv/expression.py", line 1276, in __leaf_to_sql
    params = [field.convert_to_column(right, model, validate=False)]
  File "/Users/brahoo/PycharmProjects/odoo/odoo/fields.py", line 1507, in convert_to_column
    return pycompat.to_text(value)[:self.size]
  File "/Users/brahoo/PycharmProjects/odoo/odoo/tools/pycompat.py", line 46, in to_text
    return source.decode('utf-8')
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xed in position 0: invalid continuation byte
```

File: openapi/controllers/apijsonrequest.py (hunk 1 above)
Comment (CasVissers-360ERP) [reply openapi/controllers/apijsonrequest.py:43]: > In Python 3, reading files in r mode means decoding the data into Unicode and getting a str object. Reading files in rb mode means reading the data as is, with no implicit decoding, and saving it as bytes.

I think literal strings are unicode by default in Python3?

**Comment by CasVissers** [comment]
@yelizariev 
This error:
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xed in position 0: invalid continuation byte

Seems not fixed yet, I'm still not able to run the patch method. Keep getting either the UTF-8 or Function declared as capable of handling request of type 'apijson' but called with a request of type 'json'"

Can someone maybe assist?

File: openapi/controllers/apijsonrequest.py (hunk 1 above)
Comment (yelizariev) [reply openapi/controllers/apijsonrequest.py:43]: My question was more about the variable ``httprequest.form``. Do you know anything about it? I cannot find an explanation why it should have `'r'` or `'rb'` item

Comment (CasVissers-360ERP) [reply openapi/controllers/apijsonrequest.py:43]: No I don't. I just fixed errors we ran into while trying to setup the module.

**Review by yelizariev — CHANGES_REQUESTED** [review]
Updates look reasonable

**Review by yelizariev — APPROVED** [review]
Sorry, I meant "Approved"

**Comment by CasVissers-360ERP** [comment]
-- Moved to issue.

**Pull request closed without merging**